### PR TITLE
Support JSON output for the `emp ps` command

### PR DESCRIPTION
**Pull request by mhahn**
Fixes: https://github.com/remind101/empire/issues/823

Some sample output:

```
[mhahn:empire(issue-823)]$ ./build/emp help ps
Usage: emp ps [-a <app or remote>] [-o human,json] [<name>...]

Lists processes. Shows the name, size, state, age, and command.

Examples:

    $ emp ps
    run.3794  2X  up   1m  bash
    web.1     1X  up  15h  "blog /app /tmp/dst"
    web.2     1X  up   8h  "blog /app /tmp/dst"

    $ emp ps web
    web.1     1X  up  15h  "blog /app /tmp/dst"
    web.2     1X  up   8h  "blog /app /tmp/dst"

[mhahn:empire(issue-823)]$ ./build/emp ps -a acme-inc
v1.web.dfb422e5-5187-4a73-9ff1-987c66486451     1X  RUNNING  24h  "acme-inc server"
v1.worker.80f8d14e-f0c2-4bfe-968f-fdf0a3d49bee  1X  RUNNING  24h  "acme-inc worker"

[mhahn:empire(issue-823)]$ ./build/emp ps -a acme-inc -o json
{"Processes":[{"Name":"v1.web.dfb422e5-5187-4a73-9ff1-987c66486451","Size":"1X","State":"RUNNING","Age":"24h","Command":"acme-inc server"},{"Name":"v1.worker.80f8d14e-f0c2-4bfe-968f-fdf0a3d49bee","Size":"1X","State":"RUNNING","Age":"24h","Command":"acme-inc worker"}]}

[mhahn:empire(issue-823)]$ ./build/emp ps -a acme-inc -o bla
error: error writing output: output type 'bla' is not supported```

We should be able to use this setup to easily add JSON output to other commands as well.

### CODE REVIEW

**Comment by mhahn** [comment]
Fixes https://github.com/remind101/empire/issues/823

**Comment by ejholmes** [comment]
Just to play devils advocate, we can already kinda do this with the `emp api` subcommand:

```console
$ emp api get /app/acme-inc/dynos | jq '.'
[
  {
    "attach_url": null,
    "command": "acme-inc server",
    "created_at": "0001-01-01T00:00:00Z",
    "id": "",
    "name": "v17.web.4c9dad6d-7de8-44d0-bb71-e8bf9debcaef",
    "release": {
      "id": "",
      "version": 0
    },
    "size": "256:1.00gb",
    "state": "RUNNING",
    "type": "web",
    "updated_at": "2016-06-15T04:40:12Z"
  },
  {
    "attach_url": null,
    "command": "acme-inc server",
    "created_at": "0001-01-01T00:00:00Z",
    "id": "",
    "name": "v17.web.4dfa565c-6e2a-418c-a428-8e7373ed0878",
    "release": {
      "id": "",
      "version": 0
    },
    "size": "256:1.00gb",
    "state": "RUNNING",
    "type": "web",
    "updated_at": "2016-06-14T02:46:04Z"
  }
]
```

There's definitely downsides to that, because it's coupled to the Heroku platform api types.

File: cmd/emp/main.go [hunk 1]
```diff
@@ -103,6 +109,34 @@ func (c *Command) ShortExtra() string {
...
+
+func (t *Output) JSONOutput(o interface{}) error {
+	b, err := json.Marshal(o)
+	if err == nil {
+		fmt.Println(string(b))
+	}
+	return err
+}
+
+type CommandOutput interface {
```
Comment (ejholmes) [inline cmd/emp/main.go]: I would probably make these take an io.Writer as the first argument, which they should write their output to. Also makes it easier to test the outputs, if we ever need to:

```go
type CommandOutput interface {
        WriteJSON(io.Writer) error
        WriteHuman(io.Writer) error
}
```

Or something like that.

Could also use interface upgrades to check if an output supports a certain format:

```go
func (c *Command) WriteOutput(v interface{}) error {
        w := os.Stdout

        switch ot {
        case "json":
                return json.NewEncoder(w).Encode(v)
        case "human":
                if w, ok := v.(HumanOutput); ok {
                        return w.WriteHuman(w)
                }
        }
}
```

**Comment by mhahn** [comment]
Going to close this, we should just use that `emp api` command.

**Pull request closed without merging**